ext/drx/drx.c

```diff
@@ -2614,8 +2643,192 @@ static bool
+            opnd_t dst0 = instr_get_dst(&params->inst, 0);
+            opnd_t src0 = instr_get_src(&params->inst, 0);
+            opnd_t src1 = instr_get_src(&params->inst, 1);
+            if (opnd_is_reg(dst0) && opnd_is_reg(src0) && opnd_is_reg(src1)) {
+                reg_id_t reg_dst0 = opnd_get_reg(dst0);
+                reg_id_t reg_src0 = opnd_get_reg(src0);
+                reg_id_t reg_src1 = opnd_get_reg(src1);
+                ASSERT(reg_is_gpr(reg_dst0) && reg_is_gpr(reg_src0) &&
+                           reg_is_gpr(reg_src1),
+                       "internal error: unexpected instruction format");
+                if (reg_dst0 == reg_src0 && reg_src0 == reg_src1) {
+                    params->gpr_bit_mask = reg_dst0;
+                    advance_state(DRX_DETECT_RESTORE_AVX2_GATHER_EVENT_STATE_5, params);
+                }
+            }
+        }
+        skip_unknown_instr_inc(DRX_DETECT_RESTORE_AVX512_GATHER_EVENT_STATE_0, params);
+        break;
+    case DRX_DETECT_RESTORE_AVX2_GATHER_EVENT_STATE_5:
+        if (instr_get_opcode(&params->inst) == OP_vextracti128) {
+            opnd_t src0 = instr_get_src(&params->inst, 0);
+            if (opnd_is_reg(src0)) {
+                if (opnd_get_reg(src0) == params->sg_info->mask_reg) {
+                    opnd_t dst0 = instr_get_dst(&params->inst, 0);
+                    if (opnd_is_reg(dst0)) {
+                        reg_id_t tmp_reg = opnd_get_reg(dst0);
+                        if (!reg_is_strictly_xmm(tmp_reg))
+                            break;
+                        params->the_scratch_xmm = tmp_reg;
+                        advance_state(DRX_DETECT_RESTORE_AVX2_GATHER_EVENT_STATE_6,
+                                      params);
+                        break;
+                    }
+                }
+            }
+        }
+        skip_unknown_instr_inc(DRX_DETECT_RESTORE_AVX512_GATHER_EVENT_STATE_0, params);
+        break;
+    case DRX_DETECT_RESTORE_AVX2_GATHER_EVENT_STATE_6:
+        ASSERT(params->the_scratch_xmm != DR_REG_NULL,
+               "internal error: expected xmm register to be recorded in state "
+               "machine.");
+        if ((params->sg_info->scalar_value_size == OPSZ_4 &&
+             instr_get_opcode(&params->inst) == OP_vpinsrd) ||
+            (params->sg_info->scalar_value_size == OPSZ_8 &&
+             instr_get_opcode(&params->inst) == OP_vpinsrq)) {
+            opnd_t src1 = instr_get_src(&params->inst, 1);
+            if (opnd_is_reg(src1)) {
+                if (opnd_get_reg(src1) == params->gpr_bit_mask) {
+                    ASSERT(opnd_is_reg(instr_get_dst(&params->inst, 0)),
+                           "internal error: unexpected instruction format");
+                    reg_id_t tmp_reg = opnd_get_reg(instr_get_dst(&params->inst, 0));
+                    if (tmp_reg == params->the_scratch_xmm) {
+                        advance_state(DRX_DETECT_RESTORE_AVX2_GATHER_EVENT_STATE_7,
```

Comment:
If advance_state() in every case increments by 1, you could just remove the 1st parameter and have it do the increment, removing a source of errors here where you have to manually get the next state right (imagining copy-paste mistakes...)?

Reply:
I intentionally didn't do that to get the numeric meaning off the state, in case there are state transitions > 1.